rust/lance-file/src/v2/reader.rs

```diff
@@ -218,23 +218,34 @@ impl ReaderProjection {
     ///
     /// If the schema provided is not the schema of the entire file then
     /// the projection will be invalid and the read will fail.
+    /// If the field is a `struct datatype` with `packed` set to true in the field metadata,
+    /// the whole struct has one column index.
+    /// To support nested `packed-struct encoding`, this method need to be further adjusted.
     pub fn from_whole_schema(schema: &Schema, version: LanceFileVersion) -> Self {
         let schema = Arc::new(schema.clone());
         let is_structural = version >= LanceFileVersion::V2_1;
-        let mut counter = 0;
-        let counter = &mut counter;
-        let column_indices = schema
-            .fields_pre_order()
-            .filter_map(|field| {
-                if field.children.is_empty() || !is_structural {
-                    let col_idx = *counter;
-                    *counter += 1;
-                    Some(col_idx)
-                } else {
-                    None
-                }
-            })
-            .collect::<Vec<_>>();
+        let mut column_indices = vec![];
+        let mut curr_column_idx = 0;
+        let mut packed_struct_fields_num = 0;
+        for field in schema.fields_pre_order() {
+            let field_metadata = &field.metadata;
+            if packed_struct_fields_num > 0 {
+                packed_struct_fields_num -= 1;
+                continue;
+            }
+            if field_metadata
+                .get("packed")
+                .map(|v| v == "true")
+                .unwrap_or(false)
+            {
+                column_indices.push(curr_column_idx);
+                curr_column_idx += 1;
+                packed_struct_fields_num = field.children.len();
```

Comment:
This won't be correct on nested structs.  For example:

```
A-+  
| |  
| |  
B D-+
| | |
| | |
C E F
```

If A is packed then `field.children.len()` will be 2 but there are 5 elements that need to be skipped.  I think it would be easiest to just do our own DFS here.

```
    pub fn from_whole_schema(schema: &Schema, version: LanceFileVersion) -> Self {
        let schema = Arc::new(schema.clone());
        let is_structural = version >= LanceFileVersion::V2_1;
        let mut counter = 0;
        let mut column_indices = Vec::new();
        for field in schema.fields {
            from_whole_field(field, &mut counter, &mut column_indices);
        }
        Self {
            schema,
            column_indices,
        }
    }
    fn from_whole_field(field: &Field, counter: &mut u32, column_indices: &mut Vec<u32>) {
        if is_packed_struct(field) {
             add column to column indices
             return early
        }
        if is_primitive or !is_structural {
             add column to column indices
        }
        for child in field.children {
             recurse
        }
    }
```

Reply:
nested packed struct is not supported for now, in fact, the `packed struct encoding` will panic if encounter nested `packed struct` so currently in 2.1 nested packed struct data can not be written.

inside PrimitiveStructuralEncoder::do_flush:
```
                // if the `data_block` is a `StructDataBlock`, then this is a struct with packed struct encoding.
                if let DataBlock::Struct(ref struct_data_block) = data_block {
                    if struct_data_block
                        .children
                        .iter()
                        .any(|child| !matches!(child, DataBlock::FixedWidth(_)))
                    {
                        panic!("packed struct encoding currently only supports fixed-width fields.")
                    }
                }
```